tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/ExprOptimizer.scala

```diff
@@ -62,9 +62,20 @@ class ExprOptimizer(nameGen: UniqueNameGenerator, tracker: TransformationTracker
     case OperEx(TlaSetOper.in, mem, OperEx(TlaArithOper.dotdot, left, right)) =>
       // Transform e \in a..b into a <= e /\ e <= b.
       // (The assignments are not affected by this transformation, as they are transformed to \E t \in S: x' = t.)
-      tla
-        .and(tla.le(left, mem) ? "b", tla.le(mem, right) ? "b")
-        .typed(Map("b" -> BoolT1()), "b")
+      val b = BoolT1()
+      tla.and(tla.le(left, mem).as(b), tla.le(mem, right).as(b)).as(b)
+
+    case OperEx(TlaSetOper.in, mem, OperEx(TlaSetOper.filter, NameEx(x), set, pred)) =>
+      // Transform x \in { y \in S: P } into x \in S /\ P[y/x]
+      def isToReplace: TlaEx => Boolean = {
+        case NameEx(name) => name == x
+        case _            => false
+      }
+
+      val memCopy = DeepCopy(tracker).deepCopyEx(mem)
+      val predSubstituted = ReplaceFixed(tracker)(isToReplace, memCopy)(pred)
```

Comment:
```suggestion
    case OperEx(TlaSetOper.in, mem, OperEx(TlaSetOper.filter, nameEx@NameEx(x), set, pred)) =>
      // Transform x \in { y \in S: P } into x \in S /\ P[y/x]

      def memCopy = DeepCopy(tracker).deepCopyEx(mem)
      val predSubstituted = ReplaceFixed(tracker)(nameEx, memCopy)(pred)
```

def, so that ReplaceFixed generates fresh UIDs.